# Python (programowanie funkcyjne) - funkcje
_Mikołaj Leszczuk_
<!--![](https://memegenerator.net/img/instances/68550956.jpg)-->
![](https://i.creativecommons.org/l/by/4.0/88x31.png)

## Konspekt

* **Programowanie funkcyjne** jest paradygmatem programowania, gdzie pierwsze skrzypce należą do funkcji
* W tym module poznamy podstawy programowania w tej technice w języku Python

## Funkcje anonimowe (lambda)

* Python umożliwia tworzenie funkcji w miejscu, bez osobnego deklarowania ich
* Takie funkcje nazywane są „anonimowe”, bo:
  * Nie muszą przyjmować nazw, a
  * Wykorzystuje się je często jako np. argumenty do „normalnych” funkcji
* Podstawowa składania funkcji anonimowej jest następująca:

```
lambda arg1, arg2: arg1 ** arg2
       \--------/  \----------/
        argumenty   operacja, której wynik zwraca funkcja
```

#### Podobieństwo do Excela

* **Operacja funkcji `lambda` jest niemal identyczna jak zawartość funkcji w komórce w Excelu**
* **Różnicą jest jedynie odwoływanie do:**
  * **Zmiennych (Python), i**
  * **Nazw komórek (Excel)**

#### Przykład funkcji anonimowej

* Przykład funkcji anonimowej obliczającej przeciwprostokątną w trójkącie prostokątnym (twierdzenie Pitagorasa)
* Przypomnienie:
  * Twierdzenie Pitagorasa: `c*c = a*a + b*b`
  * Operator `**` to podnoszenie do potęgi
  * Jeśli wykładnik potęgi ma postać `1/n` (przykładowo: `1/2`), to takie potęgowanie zamienia się w pierwiastkowanie o stopniu `n`

```python
lambda a, b:  ((a * a) + (b * b)) ** 0.5
```

* Funkcje anonimowe są obiektami w Pythonie
* Dzięki temu możemy je przypisać w taki sam sposób, jak przypisanie wartości do zmiennej

```python
pitagoras = lambda a, b: ((a * a) + (b * b)) ** 0.5
```

* Pozwala to w dalszym kodzie np. wywoływać taką funkcję anonimową po jej, de facto, nazwie:

In [1]:
pitagoras = lambda a, b:  ((a * a) + (b * b)) ** 0.5
print(pitagoras(3, 4))

5.0


#### Funkcje konwencjonalne vs. lambda

* Funkcja `lambda` może wyglądać jak zabawka dla matematyka 
* Formalizm, który zamienia łatwą do zrozumienia kwestię w abstrakcyjny, trudniejszy do uchwycenia formalizm
* Przede wszystkim mogliśmy osiągnąć ten sam efekt, używając po prostu następującej definicji funkcji konwencjonalnej

In [2]:
def c_sum(x, y):
    return x + y

In [3]:
l_sum = lambda x, y: x + y

In [4]:
print(c_sum(3, 4))
print(l_sum(3, 4))

7
7


* Mogę Was zapewnić, że zalety tego podejścia (`lambda`) będą widoczne, gdy nauczycie się (później) używać funkcji `map()`

## Filtrowanie, aplikowanie, redukowanie

* W Pythonie istnieje kilka przydatnych funkcji korzystających z funkcji anonimowe i pracujących na zbiorach danych:
  * `filter(funkcja, iterator)`: zwraca tylko te wartości ze zbioru danych, dla których `funkcja(wartość)` zwróciła `True`
  * `map(funkcja, iterator)`: aplikuje funkcję dla każdej wartości
  * `reduce(funkcja, iterator)`: służy do zredukowania całego zbioru danych do jednej wartości

### Filtrowanie

* Dla przykładów użyjemy prostego zbioru danych temperatury:

In [5]:
temperatury = [
    37.6, 35.8, 37.6, 33.4, 34.1, 37.1, 35.9, 34.1, 37.1, 40.5, 38.5, 37.6,
    35.8, 34.5, 36.4, 38.3, 37.5, 37.7, 34.0, 35.3, 35.7, 38.9, 34.8, 34.1,
    39.6, 35.4, 34.7, 37.6, 38.4, 36.4, 39.8, 39.1, 37.1, 35.6, 36.8, 37.6,
    36.7, 40.0, 38.0, 34.1, 35.5, 38.5, 36.1, 32.6, 32.9, 34.5, 41.0, 38.3,
    33.7, 38.7, 36.9, 36.2, 33.7, 38.3, 35.3, 38.3, 40.1, 39.3, 38.2, 37.6,
    39.1, 37.1, 34.4, 38.7, 35.8, 38.2, 38.2, 33.1, 37.8, 36.5, 37.6, 37.4,
    34.3, 37.7, 36.0, 37.5, 37.6, 36.5, 31.3, 37.7, 40.3, 39.5, 35.7, 38.1,
    34.7, 36.5, 34.3, 38.0, 37.0, 38.5, 39.4, 37.6, 41.7, 40.0, 38.4, 38.9,
    34.2, 40.2, 34.3, 35.3
]

* Użyjmy teraz funkcji filtrującej, żeby znaleźć wszystkie wartości większe lub równe `40.0`:

In [6]:
wynik = list(filter(lambda x: x >= 40.0, temperatury))
print(wynik)
wynik_sort = sorted(wynik)
print(wynik_sort)

[40.5, 40.0, 41.0, 40.1, 40.3, 41.7, 40.0, 40.2]
[40.0, 40.0, 40.1, 40.2, 40.3, 40.5, 41.0, 41.7]


In [7]:
ludzie = {'Jan': 40, 'Maria': 20}
print(list(filter(lambda x: ludzie[x] < 30, ludzie)))

['Maria']


### Aplikowanie

In [8]:
print(temperatury)

[37.6, 35.8, 37.6, 33.4, 34.1, 37.1, 35.9, 34.1, 37.1, 40.5, 38.5, 37.6, 35.8, 34.5, 36.4, 38.3, 37.5, 37.7, 34.0, 35.3, 35.7, 38.9, 34.8, 34.1, 39.6, 35.4, 34.7, 37.6, 38.4, 36.4, 39.8, 39.1, 37.1, 35.6, 36.8, 37.6, 36.7, 40.0, 38.0, 34.1, 35.5, 38.5, 36.1, 32.6, 32.9, 34.5, 41.0, 38.3, 33.7, 38.7, 36.9, 36.2, 33.7, 38.3, 35.3, 38.3, 40.1, 39.3, 38.2, 37.6, 39.1, 37.1, 34.4, 38.7, 35.8, 38.2, 38.2, 33.1, 37.8, 36.5, 37.6, 37.4, 34.3, 37.7, 36.0, 37.5, 37.6, 36.5, 31.3, 37.7, 40.3, 39.5, 35.7, 38.1, 34.7, 36.5, 34.3, 38.0, 37.0, 38.5, 39.4, 37.6, 41.7, 40.0, 38.4, 38.9, 34.2, 40.2, 34.3, 35.3]


In [9]:
from statistics import mean
sr_temp = mean(temperatury)
print(sr_temp)

36.9


In [10]:
odch = list(map(lambda x: x - sr_temp, temperatury))
print(odch)

[0.7000000000000028, -1.1000000000000014, 0.7000000000000028, -3.5, -2.799999999999997, 0.20000000000000284, -1.0, -2.799999999999997, 0.20000000000000284, 3.6000000000000014, 1.6000000000000014, 0.7000000000000028, -1.1000000000000014, -2.3999999999999986, -0.5, 1.3999999999999986, 0.6000000000000014, 0.8000000000000043, -2.8999999999999986, -1.6000000000000014, -1.1999999999999957, 2.0, -2.1000000000000014, -2.799999999999997, 2.700000000000003, -1.5, -2.1999999999999957, 0.7000000000000028, 1.5, -0.5, 2.8999999999999986, 2.200000000000003, 0.20000000000000284, -1.2999999999999972, -0.10000000000000142, 0.7000000000000028, -0.19999999999999574, 3.1000000000000014, 1.1000000000000014, -2.799999999999997, -1.3999999999999986, 1.6000000000000014, -0.7999999999999972, -4.299999999999997, -4.0, -2.3999999999999986, 4.100000000000001, 1.3999999999999986, -3.1999999999999957, 1.8000000000000043, 0.0, -0.6999999999999957, -3.1999999999999957, 1.3999999999999986, -1.6000000000000014, 1.399999

#### Format liczb zmiennoprzecinkowych

* **Podczas testów powyższego kodu, Python potrafił zamiast wartości `0.7` zwrócić `0.7000000000000028`**
* **Dlatego dla uproszczonego zapisu wszystkie wartości warto zaokrąglić do jednego miejsca po przecinku**

In [15]:
odch = list(map(lambda x: round(x - sr_temp, 1), temperatury))
print(odch)

[0.7, -1.1, 0.7, -3.5, -2.8, 0.2, -1.0, -2.8, 0.2, 3.6, 1.6, 0.7, -1.1, -2.4, -0.5, 1.4, 0.6, 0.8, -2.9, -1.6, -1.2, 2.0, -2.1, -2.8, 2.7, -1.5, -2.2, 0.7, 1.5, -0.5, 2.9, 2.2, 0.2, -1.3, -0.1, 0.7, -0.2, 3.1, 1.1, -2.8, -1.4, 1.6, -0.8, -4.3, -4.0, -2.4, 4.1, 1.4, -3.2, 1.8, 0.0, -0.7, -3.2, 1.4, -1.6, 1.4, 3.2, 2.4, 1.3, 0.7, 2.2, 0.2, -2.5, 1.8, -1.1, 1.3, 1.3, -3.8, 0.9, -0.4, 0.7, 0.5, -2.6, 0.8, -0.9, 0.6, 0.7, -0.4, -5.6, 0.8, 3.4, 2.6, -1.2, 1.2, -2.2, -0.4, -2.6, 1.1, 0.1, 1.6, 2.5, 0.7, 4.8, 3.1, 1.5, 2.0, -2.7, 3.3, -2.6, -1.6]


### Redukowanie

* Funkcja `reduce` z modułu `functools` pomaga w iteratywnym redukowaniu zbioru wartości do jednej wartości
* Żeby to lepiej zobrazować, pokażmy dwa przykłady użycia `reduce` do obliczenia:
  * Sumy wartości, i
  * Ich iloczynu

#### Przykład 1: obliczenie sumy

In [16]:
from functools import reduce
nums = [1, 2, 3, 4, 5]
print(reduce(lambda a, b: a + b, nums))

15


* Użycie funkcji `reduce` w powyższym przykładzie sprowadza się tak naprawdę do:

In [17]:
print((((1 + 2) + 3) + 4) + 5)

15


#### Przykład 2: obliczenie iloczynu

In [18]:
print(reduce(lambda a, b: a * b, [1, 2, 3, 4]))

24


* Sprowadza się to do:

In [19]:
print(((1 * 2) * 3) * 4)

24


## Tworzenie zbiorów danych w locie

### Tworzenie list – ang. *list comprehension*

* Dzięki generowaniu list w locie możemy:
  * Nawet dość skomplikowane pętle, zamienić na
  * Pojedyncze linijki kodu

* Przykładowo 3 linijki kodu generujące listę sześcianów:

In [20]:
szesciany = []
for x in range(10):
    szesciany.append(x**3)

In [21]:
print(szesciany)

[0, 1, 8, 27, 64, 125, 216, 343, 512, 729]


* W podstawowej wersji możemy „przenieść” pętlę do nawiasów kwadratowych:

In [23]:
szesciany = [x**3 for x in range(10)]

In [24]:
print(szesciany)

[0, 1, 8, 27, 64, 125, 216, 343, 512, 729]


* Ogólniejsza postać *list comprehension*:

```python
nowa_lista = [funkcja(el) for el in lista if warunek(el)]
```

* Na przykład by przygotować listę kwadratów liczb nieparzystych z zakresu od `1` do `101`:

In [25]:
kwadraty = [el**2 for el in range(1, 102) if el % 2 != 0]

In [26]:
print(kwadraty)

[1, 9, 25, 49, 81, 121, 169, 225, 289, 361, 441, 529, 625, 729, 841, 961, 1089, 1225, 1369, 1521, 1681, 1849, 2025, 2209, 2401, 2601, 2809, 3025, 3249, 3481, 3721, 3969, 4225, 4489, 4761, 5041, 5329, 5625, 5929, 6241, 6561, 6889, 7225, 7569, 7921, 8281, 8649, 9025, 9409, 9801, 10201]


### Tworzenie zbiorów

* W podobny sposób można też przygotować zbiór

In [27]:
zbior = {znak for znak in "abracadabra" if znak not in "abc"}

In [28]:
print(zbior)

{'r', 'd'}


### Tworzenie słowników

* Podobnie można stworzyć słownik

In [29]:
tekst = "abracadabra"
wystapienia = {znak: tekst.count(znak) for znak in tekst}

In [30]:
print(wystapienia)

{'a': 5, 'b': 2, 'r': 2, 'c': 1, 'd': 1}


In [31]:
c = {x: x**2 for x in (2, 4, 6)}  # {2: 4, 4: 16, 6: 36}

In [32]:
print(c)

{2: 4, 4: 16, 6: 36}


## Wyrażenia generatorowe – ang. *generator expressions*

* Co do zapisu: nie różnią się niczym od *list comprehension*, poza zmianą znaków nawiasów z `[]` na `()`

In [33]:
list_comp = [x ** 0.5 for x in range(1, 11)]
gene_expr = (x ** 0.5 for x in range(1, 11))

In [34]:
for x in list_comp:
    print(x)

1.0
1.4142135623730951
1.7320508075688772
2.0
2.23606797749979
2.449489742783178
2.6457513110645907
2.8284271247461903
3.0
3.1622776601683795


In [35]:
for x in gene_expr:
    print(x)

1.0
1.4142135623730951
1.7320508075688772
2.0
2.23606797749979
2.449489742783178
2.6457513110645907
2.8284271247461903
3.0
3.1622776601683795


* Różnica polega na tym, że *list comprehension* tworzy listę, a więc zajmuje miejsce w pamięci i czas procesora
* Wyrażenie generatorowe jest dopiero obliczane przy przechodzeniu przez nie, np. w formie pętli `for`

In [38]:
list_comp = [x ** 0.5 for x in range(1, 50_000_001)]

In [39]:
suma = 0
for x in list_comp:
    suma += x
print(suma)

235702263930.8374


In [42]:
gene_expr = (x ** 0.5 for x in range(1, 50_000_001))
# gene_expr nie obliczyło jeszcze tych milionów pierwiastków

In [43]:
suma = 0
for x in gene_expr:
    # teraz już kolejne elementy wyrażenia generatorowego są obliczane
    suma += x
print(suma)

235702263930.8374


## Ćwiczenia

### Ćwiczenie na rozgrzewkę

* Utwórz funkcję `lambda`, która przyjmuje jeden parametr (`a`) i zwraca go

### Rozwiązanie ćwiczenia na rozgrzewkę

In [44]:
x = lambda a : a

In [45]:
print(x(3))

3


### Kolejne ćwiczenie na rozgrzewkę

Napisz program w Pythonie, aby utworzyć funkcję `lambda`, która dodaje `15` do podanej liczby przekazanej jako argument i zwraca to jako wynik. Następnie wydrukuj wynik.

### Rozwiązania kolejnego ćwiczenia na rozgrzewkę

In [48]:
r = lambda a: a + 15
print(r(10))

25


### Kolejne ćwiczenie na rozgrzewkę

Utwórz funkcję `lambda`, która mnoży argument `x` przez argument `y` i zwraca to jako wynik. Następnie wydrukuj wynik.

### Rozwiązania kolejnego ćwiczenia na rozgrzewkę

In [49]:
r = lambda x, y: x * y
print(r(12, 4))

48


### Ćwiczenie

Napisz program w Pythonie wsparcia sortowania listy krotek za pomocą `lambda` po drugim elemencie

```python
subject_marks = [('Język angielski', 88),
                 ('Nauka',           90),
                 ('Matematyka',      97),
                 ('Nauki społeczne', 82)]
```

### Przykładowe rozwiązanie

In [50]:
subject_marks = [('Język angielski', 88),
                 ('Nauka',           90),
                 ('Matematyka',      97),
                 ('Nauki społeczne', 82)]
print("Oryginalna lista krotek:")
print(subject_marks)
subject_marks.sort(key=lambda x: x[1])
print("\nSortowanie listy krotek:")
print(subject_marks)

Oryginalna lista krotek:
[('Język angielski', 88), ('Nauka', 90), ('Matematyka', 97), ('Nauki społeczne', 82)]

Sortowanie listy krotek:
[('Nauki społeczne', 82), ('Język angielski', 88), ('Nauka', 90), ('Matematyka', 97)]


### Ćwiczenie

Napisz program w Pythonie, aby posortować za pomocą `lambda` listę słowników po kluczu `model` lub `kolor`.

```python
models = [{'marka': 'Nokia',   'model': '3310',   'kolor': 'Czarny'},
          {'marka': 'Apple',   'model': '11',     'kolor': 'Złoty'},
          {'marka': 'Samsung', 'model': 'Galaxy', 'kolor': 'Srebrny'}]
```

### Przykładowe rozwiązanie

In [52]:
models = [{'marka': 'Nokia',   'model': '3310',   'kolor': 'Czarny'},
          {'marka': 'Apple',   'model': '11',     'kolor': 'Złoty'},
          {'marka': 'Samsung', 'model': 'Galaxy', 'kolor': 'Srebrny'}]
print("Oryginalna lista słowników:")
print(models)
sorted_models = sorted(models, key=lambda x: x['model'])
print("\nSortowanie listy słowników:")
print(sorted_models)

Oryginalna lista słowników:
[{'marka': 'Nokia', 'model': '3310', 'kolor': 'Czarny'}, {'marka': 'Apple', 'model': '11', 'kolor': 'Złoty'}, {'marka': 'Samsung', 'model': 'Galaxy', 'kolor': 'Srebrny'}]

Sortowanie listy słowników:
[{'marka': 'Apple', 'model': '11', 'kolor': 'Złoty'}, {'marka': 'Nokia', 'model': '3310', 'kolor': 'Czarny'}, {'marka': 'Samsung', 'model': 'Galaxy', 'kolor': 'Srebrny'}]


### Ćwiczenie

* Napisz program w Pythonie, aby sprawdzić, czy dany ciąg (`str`) zaczyna się od znaku '`P`', używając `lambda`
* Podpowiedź: skorzystaj z funkcji (metody) `startswith()`

### Przykładowe rozwiązanie

In [53]:
starts_with = lambda x: x.startswith('P')

In [54]:
print(starts_with('Python'))

True


In [55]:
print(starts_with('Java'))

False


### Ćwiczenie

* Napisz program w Pythonie, aby wyodrębnić rok, miesiąc, dzień i godzinę za pomocą `lambda`
* Podpowiedź: skorzystaj z modułu `datetime`:
  * `now = datetime.datetime.now()` - przypisuje do `now` aktualną lokalną datę i godzinę.
  * `now.year` - wyodrębnia i zwraca rok z `now`.
  * `now.month` - wyodrębnia i zwraca miesiąc z `now`.
  * `now.day` - wyodrębnia i zwraca dzień z `now`.
  * `now.time()` - wyodrębnia i zwraca godzinę z `now`.

### Przykładowe rozwiązanie

In [56]:
import datetime
now = datetime.datetime.now()
print(now)
year = lambda x: x.year
month = lambda x: x.month
day = lambda x: x.day
t = lambda x: x.time()
print(year(now))
print(month(now))
print(day(now))
print(t(now))

2023-09-14 20:17:12.597818
2023
9
14
20:17:12.597818


### Ćwiczenie

* Napisz program w Pythonie, aby sprawdzić, czy dany ciąg jest liczbą, czy nie, używając `lambda`
* Podpowiedź: przydatna metoda to\
    `string.replace(oldvalue, newvalue, count)`\
    Składnia parametrów:
  * `oldvalue`	– 	Wymagany; ciąg do wyszukania
  * `newvalue`	– 	Wymagany; ciąg znaków, który ma zastąpić starą wartość
  * `count`		– 	Opcjonalny; liczba określająca, ile wystąpień starej wartości chcesz zastąpić; domyślnie są to wszystkie wystąpienia

### Przykładowe rozwiązanie

In [57]:
is_num = lambda q: q.replace('.', '', 1).isdigit()
print(is_num('26587'))
print(is_num('4.2365'))
print(is_num('-12547'))
print(is_num('00'))
print(is_num('A001'))
print(is_num('001'))
print("\nWydrukuj liczby kontrolne:")
is_num1 = lambda r: is_num(r[1:]) if r[0] == '-' else is_num(r)
print(is_num1('-16.4'))
print(is_num1('-24587.11'))

True
True
False
True
False
True

Wydrukuj liczby kontrolne:
True
True


### Ćwiczenie

Napisz program w Pythonie do filtrowania listy liczb parzystych i nieparzystych całkowitych za pomocą `lambda` i `filter`

```python
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```

### Przykładowe rozwiązanie

In [58]:
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
print("Oryginalna lista liczb całkowitych:")
print(nums)
print("\nParzyste liczby ze wspomnianej listy:")
even_nums = list(filter(lambda x: x % 2 == 0, nums))
print(even_nums)
print("\nNieparzyste liczby ze wspomnianej listy:")
odd_nums = list(filter(lambda x: x % 2 != 0, nums))
print(odd_nums)

Oryginalna lista liczb całkowitych:
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

Parzyste liczby ze wspomnianej listy:
[2, 4, 6, 8, 10]

Nieparzyste liczby ze wspomnianej listy:
[1, 3, 5, 7, 9]


### Ćwiczenie

Napisz program w Pythonie, aby znaleźć przecięcie dwóch podanych list za pomocą `lambda` i `filter`

```python
array_nums1 = [1, 2, 3, 5, 7, 8, 9, 10]
array_nums2 = [1, 2, 4, 8, 9]
```

### Przykładowe rozwiązanie

In [59]:
array_nums1 = [1, 2, 3, 5, 7, 8, 9, 10]
array_nums2 = [1, 2, 4, 8, 9]
print("Oryginalne tablice:")
print(array_nums1)
print(array_nums2)
result = list(filter(lambda x: x in array_nums1, array_nums2))
print("\nPrzecięcie wspomnianych tablic:", result)

Oryginalne tablice:
[1, 2, 3, 5, 7, 8, 9, 10]
[1, 2, 4, 8, 9]

Przecięcie wspomnianych tablic: [1, 2, 8, 9]


### Ćwiczenie

Napisz program w Pythonie, aby policzyć parzyste i nieparzyste liczby w danej tablicy liczb całkowitych, używając `lambda` i `filter`

```python
array_nums = [1, 2, 3, 5, 7, 8, 9, 10]
```

### Przykładowe rozwiązanie

In [60]:
array_nums = [1, 2, 3, 5, 7, 8, 9, 10]
print("Oryginalna tablica:")
print(array_nums)
odd_ctr = len(list(filter(lambda x: x % 2 != 0, array_nums)))
even_ctr = len(list(filter(lambda x: x % 2 == 0, array_nums)))
print("\nLiczba liczb parzystych w powyższej tablicy: ", even_ctr)
print("\nLiczba liczb nieparzystych w powyższej tablicy: ", odd_ctr)

Oryginalna tablica:
[1, 2, 3, 5, 7, 8, 9, 10]

Liczba liczb parzystych w powyższej tablicy:  3

Liczba liczb nieparzystych w powyższej tablicy:  5


### Ćwiczenie

Napisz program w Pythonie, aby znaleźć wartości o długości sześć na podanej liście za pomocą funkcji `lambda` i `filter`

```python
weekdays = ['Poniedziałek', 'Wtorek', 'Środa', 'Czwartek', 'Piątek', 'Sobota', 'Niedziela']
```

### Przykładowe rozwiązanie

In [63]:
weekdays = ['Poniedziałek', 'Wtorek', 'Środa', 'Czwartek', 'Piątek', 'Sobota', 'Niedziela']
days = filter(lambda day: len(day) == 6, weekdays)
# print(days)
for d in days:
    print(d)

Wtorek
Piątek
Sobota


### Ćwiczenie

Napisz program w Pythonie, aby znaleźć liczby podzielne przez dziewiętnaście lub trzynaście z listy liczb za pomocą `lambda` i `filter`

```python
nums = [19, 65, 57, 39, 152, 639, 121, 44, 90, 190]
```

### Przykładowe rozwiązanie

In [64]:
nums = [19, 65, 57, 39, 152, 639, 121, 44, 90, 190]
print("Oryginalna lista:")
print(nums)
result = list(filter(lambda x: x % 19 == 0 or x % 13 == 0, nums))
print("\nLiczby z poniższej listy są podzielne przez 19 lub 13:")
print(result)

Oryginalna lista:
[19, 65, 57, 39, 152, 639, 121, 44, 90, 190]

Liczby z poniższej listy są podzielne przez 19 lub 13:
[19, 65, 57, 39, 152, 190]


### Ćwiczenie

* Napisz program w Pythonie, aby znaleźć palindromy na podanej liście ciągów za pomocą `lambda` i `filter`
* Palindrom – wyrażenie brzmiące tak samo czytane od lewej do prawej i od prawej do lewej
* Przykładem palindromu jest: „kobyła ma mały bok”

```python
texts = ["php", "w3r", "Python", "abcd", "Java", "aaa"]
```

### Przykładowe rozwiązanie

In [65]:
texts = ["php", "w3r", "Python", "abcd", "Java", "aaa"]
print("Oryginalna lista ciągów:")
print(texts)
result = list(filter(lambda x: x == x[::-1], texts))
print("\nLista palindromów:")
print(result)

Oryginalna lista ciągów:
['php', 'w3r', 'Python', 'abcd', 'Java', 'aaa']

Lista palindromów:
['php', 'aaa']


### Ćwiczenie

* Napisz program w Pythonie, który zsumuje długość imion z danej listy imion, ale tylko takich, które rozpoczynają się od dużej litery, a pozostałe litery mają małe
* Użyj funkcji `lambda`

```python
sample_names = ['antoni', 'Jakub', 'zuzanna', 'Julia', 'Jan', 'szymon']
```

### Przykładowe rozwiązanie

In [66]:
sample_names = ['antoni', 'Jakub', 'zuzanna', 'Julia', 'Jan', 'szymon']
sample_names=list(filter(lambda el: el[0].isupper() and el[1:].islower(), sample_names))
print("Wynik:")
print(len(''.join(sample_names)))

Wynik:
13


### Ćwiczenie

Napisz program w Pythonie podnoszący do kwadratu i sześcianu każdą liczbę z podanej listy liczb całkowitych, używając `lambda` i `map`

```python
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```

### Przykładowe rozwiązanie

In [ ]:
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
print("Oryginalna lista liczb całkowitych:")
print(nums)
print("\nKwadrat każdej liczby z listy:")
square_nums = list(map(lambda x: x ** 2, nums))
print(square_nums)
print("\nSześcian każdej liczby z listy:")
cube_nums = list(map(lambda x: x ** 3, nums))
print(cube_nums)

### Ćwiczenie

* Napisz program w Pythonie, który za pomocą funkcji `lambda` mnoży każdą liczbę z podanej listy przez określoną liczbę
* Wydrukuj wynik

```python
nums = [2, 4, 6, 9, 11]
n = 2
```

### Przykładowe rozwiązanie

In [ ]:
nums = [2, 4, 6, 9, 11]
n = 2
print("Oryginalna lista:", nums)
print("Podana liczba:", n)
filtered_numbers = list(map(lambda number: number*n, nums))
print("Wynik jako list:", filtered_numbers)
filtered_numbers_spaces = '  '.join(map(str, filtered_numbers))
print("Wynik jako str:", filtered_numbers_spaces)

### Ćwiczenie

Napisz program w Pythonie, aby dodać dwie podane listy za pomocą `map` i `lambda`

```python
nums1 = [1, 2, 3]
nums2 = [4, 5, 6]
```

### Przykładowe rozwiązanie

In [ ]:
nums1 = [1, 2, 3]
nums2 = [4, 5, 6]
print("Oryginalna lista:")
print(nums1)
print(nums2)
result = map(lambda x, y: x + y, nums1, nums2)
print("\nWynik: po dodaniu dwóch list")
print(list(result))

### Ćwiczenie

* Napisz program w Pythonie, który usuwa liczby dodatnie z podanej listy liczb
* Zsumuj liczby ujemne i wydrukuj wartość bezwzględną za pomocą tworzenia listy – ang. *list comprehension*
* Wydrukuj wynik

```python
nums = [2, 4, -6, -9, 11, -12, 14, -5, 17]
```

### Przykładowe rozwiązanie

In [ ]:
nums = [2, 4, -6, -9, 11, -12, 14, -5, 17]
print("Oryginalna lista: ", nums)
print("Wynik:")
print(abs(sum([i for i in nums if i < 0])))

### Ćwiczenie

* Napisz program w Pythonie, aby zmienić kolejność liczb dodatnich i ujemnych w danej liście (najpierw wszystkie ujemne, potem wszystkie dodatnie) za pomocą tworzenia listy – ang. *list comprehension*

```python
array_nums = [-1, 2, -3, 5, 7, 8, 9, -10]
```

### Przykładowe rozwiązanie

In [ ]:
array_nums = [-1, 2, -3, 5, 7, 8, 9, -10]
print("Oryginalna tablica:")
print(array_nums)
result = [x for x in array_nums if x < 0] + [x for x in array_nums if x >= 0]
print("\nZmiana kolejności liczb dodatnich i ujemnych wspomnianej tablicy:")
print(result)

### Ćwiczenie

* Napisz program w Pythonie, aby:
  * Znaleźć liczby z podanego ciągu
  * Zapisać je na liście
  * Wyświetlić liczby w posortowanej formie
* Użyj funkcji tworzenia listy – ang. *list comprehension*, aby rozwiązać problem

```python
str1 = "sdf 23 safs8 5 sdfsd8 sdfs 56 21sfs 20 5"
```

### Przykładowe rozwiązanie

In [ ]:
str1 = "sdf 23 safs8 5 sdfsd8 sdfs 56 21sfs 20 5"
print("Oryginalny ciąg: ", str1)
str_num = str1.split()
print("Oryginalna lista: ", str_num)
numbers = sorted([int(x) for x in str_num if x.isdigit()])
print('Liczby w posortowanej formie:')
for i in numbers:
    print(i, end=' ')

## Zadania utrwalające

### Zadanie 1

* Napisz funkcję anonimową obliczającą kwadrat danej wartości
* Czy potrafisz wytłumaczyć czym się to różni od użycia funkcji
```python
pow(a, b)
```
lub operatora
```python
a ** b
```
?

### Zadanie 2

* Napisz funkcję anonimową obliczającą średnią arytmetyczną dwóch wartości

### Zadanie 3

* Napisz funkcję anonimową sprawdzającą czy dana wartość jest większa (lub mniejsza) od pewnej wartości
* Taka funkcja anonimowa powinna zwracać wartość logiczną

### Zadanie 4

* Znajdź wszystkie wartości mniejsze od `36.6`

### Zadanie 5

* Odfiltruj wszystkie wartości, które są:
  * Mniejsze bądź równe `35.0`, lub 
  * Większe bądź równe `40.0`

### Zadanie 6

* Oblicz dla każdej liczby w zbiorze temperatur kwadrat:
  * Różnicy tej liczby, i
  * Średniej arytmetycznej całego zbioru

### Zadanie 7

* Mając obliczone wartości odchyleń temperatur od średniej temperatury, oblicz **wariancję** tych wartości
* Wariancja:
  * Klasyczna **miara zmienności**
  * Intuicyjnie utożsamiana ze zróżnicowaniem zbiorowości
  * Jest **średnią arytmetyczną kwadratów odchyleń** (różnic):
    * Poszczególnych wartości **cechy**, od 
    * **Wartości oczekiwanej**
  * [https://www.matemaks.pl/wariancja.html](https://www.matemaks.pl/wariancja.html)
* Nie zapomnij o podzieleniu przez ilość elementów w zbiorze!

### Zadanie 8

* Napisz pogram, który policzy kwadraty liczb z zakresu `[1,10000]`, które podzielne są przez:
  * `5`, lub 
  * `9`
* Następnie sprawdź, które z uzyskanych liczb są podzielne:
  * Zarówno przez `5`,
  * Jak i przez `9`

### Zadanie 9

* Stwórz zbiór, który wypełnisz `30` wywołaniami funkcji:
```python
random.randint(1, 30)
```
* Ile razy wylosowana została ta sama wartość?
* Sprawdzisz to porównując liczbę elementów zbioru.
* Jak się zachowa program, jeśli zmienisz argumenty do funkcji `randint`? 
* A jeśli zmienisz liczbę wywołań tej funkcji?

### Zadanie 10

* Napisz program w Pythonie do tworzenia ciągu Fibonacciego aż do `n` za pomocą `lambda` i `reduce`
* Podpowiedzi:
  * Da się to zrobić w zasadzie w jednej linijce!
  * Przypomnij sobie, co robiła zmienna `_`

### Przykładowe rozwiązanie zadania 10

In [ ]:
from functools import reduce
fib_series = lambda n: reduce(lambda x, _: x + [x[-1] + x[-2]],range(n - 2), [0, 1])

In [ ]:
print("Ciąg Fibonacciego do 2:")
print(fib_series(2))

In [ ]:
print("Ciąg Fibonacciego do 5:")
print(fib_series(5))

In [ ]:
print("Ciąg Fibonacciego do 6:")
print(fib_series(6))

In [ ]:
print("Ciąg Fibonacciego do 9:")
print(fib_series(9))